### ISSUE TITLE
Create different Biomes

### ISSUE THREAD

**Issue by IPotatoCouch** [Eldemarkki/Marching-Cubes-Terrain#40]
Different biomes / regions would be nice to have :]

**Comment by Eldemarkki** [comment]
This solution that I quickly thought up of would only work with 2 biomes, but maybe someone smarter than me knows how to extend it to more than 2. I haven't spent much time on this.

First, create 2 noise functions; 1 for biome A called `GetA(x, z)`, and one for biome B called `GetB(x, z)`. Those functions would generate the height for the biomes. So `GetA()` could for example be an infinite plains biome, and `GetB()` and infinite mountains biome.

Then, have a third noise function representing the blend. For any given point (x, z), the height of point (x, z) would be `y = Lerp(GetA(x, z), GetB(x, z), GetBlend(x, z))`. `GetA()` and `GetB()` representing the noise functions of A and B. `GetBlend()` would tell how much to blend the biomes: 0 would be completely biome A, and 1 would be completely biome B.

**Comment by IPotatoCouch** [comment]
Thanks for the tip! This sounds like a good solution. For multiple biomes I was thinking would it be possible to generate a biomes map, something like this maybe.

![BiomesMap](https://user-images.githubusercontent.com/68332920/87804679-e1db1800-c85c-11ea-91f5-a364ade318c1.png)

Each biome has its own color, so with a map like this it would be easy to determine where each biome is located. One issue is blending biomes together. Maybe even make biomes bleed into each other, so it would give more realistic results. I'm not smart enough to figure this out tho..

**Comment by Eldemarkki** [comment]
Yeah, the blend is very difficult. I remember once (like maybe a year ago) trying a climate based biome system, where the inputs are 2 noise maps, one for precipitation and one for temperature. Here's picture of what I mean:
![](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fmiddleschoolscienceblog.files.wordpress.com%2F2015%2F01%2Fbiomes_graph.jpg&f=1&nofb=1)

In that experiment, I think I could generate the biomes nicely, but blending the heights was the biggest problem which I never managed to solve.

**Comment by IPotatoCouch** [comment]
That looks really nice! What if you blend the biomes map like this? Each biome has its own noise so once the map is blended so as the noise maps. Would that be possible?

![BiomesBlend](https://user-images.githubusercontent.com/68332920/87813347-65e7cc80-c86a-11ea-8b6a-ca8cecfefbc8.png)
![BlendedBiomes](https://user-images.githubusercontent.com/68332920/87815475-566a8280-c86e-11ea-9055-8247a769bac0.png)

This is a bit ugly example, but does this even make sense? In my head this sounds logical yet I have no idea how to accomplish this.. :D

**Comment by IPotatoCouch** [comment]
Biomes map would almost act like a mask for noise maps, if the biomes map is 'blurred' you would get a smooth transition

**Comment by Eldemarkki** [comment]
I think that's a great idea! I'll post something here if I manage to think of some implementation. It's definitely a hard problem.

**Comment by IPotatoCouch** [comment]
Awesome thank you! 🙂 Been scratching my head for hours, it's quite a task.. A bit too much for my brain too handle 😄

**Comment by Eldemarkki** [comment]
I looked for the repository about the climate based biomes (like in the diagram in the previous comment) and found it! I made a few changes to fix the blending, and made the repository public. Here it is: https://github.com/Eldemarkki/TerrainBiomeGenerator. 

This is what it looks like with blending=0:
![kuva](https://user-images.githubusercontent.com/38920928/87837749-46b66280-c89d-11ea-8ab1-29c5d51cf10a.png)

And this is with blending=3:
![kuva](https://user-images.githubusercontent.com/38920928/87837763-5766d880-c89d-11ea-956b-ae88c8354180.png)

It's not perfect, the colours are quite sharp looking at some places, and it is not very realistic as there are multiple places where it goes from desert to tundra, but that can be tweaked quite easily. 

And in case it wasn't obvious, that is not using marching cubes.

The blending is done by just taking the average height around the sampling point. The transition from any biome to ocean is not blended, because I don't want the ocean to suddenly go upwards near the transition.

**Comment by IPotatoCouch** [comment]
Really nice! This works pretty well! Will you be implementing this approach to _Marching-Cubes-Terrain_ project as well? 🙂

**Comment by Eldemarkki** [comment]
I'm not really sure. If I end up implementing it, it most likely will not be in the [main repository](https://github.com/Eldemarkki/Marching-Cubes-Terrain/), but in the [TerrainBiomeGenerator repository](https://github.com/Eldemarkki/TerrainBiomeGenerator).
